### PR TITLE
RPC API

### PR DESCRIPTION
**Pull request by momintlh**
## Changes:
* Added RPCRegister
* Added RPCCall
* Added enum for RPC Modes in Unity.

### Next Tasks:
- [x] Currently only String data can be passed, maybe giving a helper function to convert data to JSON?
- [x] Callback in RPCregister from Unity. 
- [x] List Of Callbacks
- [x] Parsing data and sender
- [x] Clean-Up

## Testing:
Verified the changes by testing RPC registrations and calls with different modes from both C# and JavaScript. Checked the sequence of ENUM values and ensured correct callback invocations. 

## Test Results:
* Host registers and gets the following response:
image
> Rock is the data passed by sender / the nonhost player in this case.

* Nonhost player calls:
image
> here response (Okay!) is being received from the register callback

based on #37 

--- 

Final demo output: https://mmntlh.itch.io/playroom-demo

### CODE REVIEW

File: Assets/PlayroomKit/PlayroomKit.cs [hunk 1]
```diff
@@ -971,6 +970,41 @@ public class Dpad
...
+        // RPC:
+        public enum RPCModes
+        {
+            ALL,
+            OTHERS,
+            HOST
+        }
+
+        [DllImport("__Internal")]
+        public extern static void RPCregister(string name);
```
Comment (SaadBazaz) [inline Assets/PlayroomKit/PlayroomKit.cs:983]: Minor nit: function naming convention

Reference: https://chat.openai.com/share/ea8c8c00-eb88-48ed-ba62-0967a7b48b50

File: Assets/PlayroomKit/PlayroomKit.cs [hunk 2]
```diff
@@ -971,6 +970,41 @@ public class Dpad
...
+
+        [DllImport("__Internal")]
+        private extern static void RPCcallInternal(string name, string data, RPCModes mode, Action callbackOnResponse);
+
+        private static Action CallbackOnResponse = null;
```
Comment (SaadBazaz) [inline Assets/PlayroomKit/PlayroomKit.cs:988]: Have we tested this in the case of multiple RPC callback registrations? Wouldn't this be overridden in that case?

**Review by SaadBazaz — CHANGES_REQUESTED** [review]
Needs clarification.

File: Assets/PlayroomKit/PlayroomKit.cs (hunk 1 above)
Comment (momintlh) [reply Assets/PlayroomKit/PlayroomKit.cs:983]: My bad, will update to PascalCase for consistency.

File: Assets/PlayroomKit/PlayroomKit.cs (hunk 2 above)
Comment (momintlh) [reply Assets/PlayroomKit/PlayroomKit.cs:988]: > Have we tested this in the case of multiple RPC callback registrations? Wouldn't this be overridden in that case?

Just did a test with this and yes It does get overridden (Latest callback is called): 
image

I guess holding a list of callbacks should work?

Comment (SaadBazaz) [reply Assets/PlayroomKit/PlayroomKit.cs:988]: Perhaps yes. Also, this brings up another edge case:
1. Multiple RPC callback registrations for the **same RPC event**
2. Multiple RPC callback registrations for **different RPC events**

Both should work.

**Comment by momintlh** [comment]
The issue where callback was getting overridden is fixed, so now each RPC event can call its own callback, but when the same event has different callbacks, only the latest one gets invoked. Current solution I have is to change the ```Dictionary<string, Action>``` to ```Dictionary<string, List<Action>>```, so each event can hold multiple callbacks, But I might be Overengineering this part, so any suggestions would be helpful. @SaadBazaz @asadm.

Changes are in this commit: [c60dc8d](https://github.com/asadm/playroom-unity/pull/43/commits/c60dc8d08460cb190ab729e0c7c0a7a9923b7d8a)

**Comment by SaadBazaz** [comment]
> The issue where callback was getting overridden is fixed, so now each RPC event can call its own callback, but when the same event has different callbacks, only the latest one gets invoked. Current solution I have is to change the `Dictionary<string, Action>` to `Dictionary<string, List<Action>>`, so each event can hold multiple callbacks, But I might be Overengineering this part, so any suggestions would be helpful. @SaadBazaz @asadm.
> 
> Changes are in this commit: [c60dc8d](https://github.com/asadm/playroom-unity/pull/43/commits/c60dc8d08460cb190ab729e0c7c0a7a9923b7d8a)

I think your approach is fine. Do you have any other solutions in mind?

**Comment by asadm** [comment]
multiple callbacks for one rpc? yes.

File: Assets/Plugins/PlayroomPlugin.jslib [hunk 3]
```diff
@@ -32,13 +32,16 @@ mergeInto(LibraryManager.library, {
...
+
     var options = optionsJson ? JSON.parse(UTF8ToString(optionsJson)) : {};
 
+
     Promise.all([
       embedScript("https://unpkg.com/react@18.2.0/umd/react.development.js"),
       embedScript(
         "https://unpkg.com/react-dom@18.2.0/umd/react-dom.development.js"
       ),
+      // not the latest version:
```
Comment (SaadBazaz) [inline Assets/Plugins/PlayroomPlugin.jslib:44]: Unnecessary comment? I can't see the version in place.

File: Assets/PlayroomKit/PlayroomKit.cs [hunk 4]
```diff
@@ -971,6 +971,167 @@ public class Dpad
...
+                JSONArray arrayNode = new JSONArray();
+                foreach (object element in enumerable)
+                {
+                    arrayNode.Add(ConvertToJson(element));
+                }
+                return arrayNode.ToString();
+            }
+            else
+            {
+                Debug.Log($"{data} is '{data.GetType()}' which is currently not supported by RPC!");
```
Comment (SaadBazaz) [inline Assets/PlayroomKit/PlayroomKit.cs:1129]: Can you drop an explainer why this is necessary? Why can't we support these types?

File: Assets/PlayroomKit/PlayroomKit.cs [hunk 5]
```diff
@@ -971,6 +971,167 @@ public class Dpad
...
+            }
+            else if (data.GetType().IsPrimitive || data is string)
+            {
+                return data.ToString();
+            }
+            if (data is Vector2 vector2)
+            {
+                return JsonUtility.ToJson(vector2);
+            }
+            else if (data is Vector3 vector3)
```
Comment (SaadBazaz) [inline Assets/PlayroomKit/PlayroomKit.cs:1089]: Does this also impact #20 ?

File: Assets/PlayroomKit/PlayroomKit.cs [hunk 6]
```diff
@@ -971,6 +971,167 @@ public class Dpad
...
+            RpcRegisterCallback?.Invoke(dataJson, senderJson);
+        }
+
+        [DllImport("__Internal")]
+        private extern static void RpcCallInternal(string name, string data, RPCModes mode, Action callbackOnResponse);
+
+        private static Dictionary<string, List<Action>> OnResponseCallbacks = new Dictionary<string, List<Action>>();
+        private static List<string> RpcEventNames = new List<string>();
+
+        public static void RpcCall(string name, object data, RPCModes mode, Action callbackOnResponse)
```
Comment (SaadBazaz) [inline Assets/PlayroomKit/PlayroomKit.cs:1022]: In PlayroomKit JS, the function name is RPC.call, which implies RPC is an object (or a Class!) containing `call` as a function. Why didn't we follow that in particular?

Comment (SaadBazaz) [reply Assets/PlayroomKit/PlayroomKit.cs:1022]: (An explanation would do if this feature is not needed right now)

**Review by SaadBazaz — APPROVED** [review]
Approved with comments.

File: Assets/PlayroomKit/PlayroomKit.cs (hunk 5 above)
Comment (momintlh) [reply Assets/PlayroomKit/PlayroomKit.cs:1089]: No this is separate.

File: Assets/PlayroomKit/PlayroomKit.cs (hunk 6 above)
Comment (SaadBazaz) [reply Assets/PlayroomKit/PlayroomKit.cs:1027]: Makes sense for now. If we expand this later on, we may need to put it in a class, which would be a breaking change.

**Pull request merged**